streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2125,17 +2125,34 @@ object ZStream extends ZStreamPlatformSpecific {
    * Creates a stream from a [[zio.ZQueue]] of values
    */
   final def fromQueue[R, E, A](queue: ZQueue[_, _, R, E, _, A]): ZStream[R, E, A] =
-    unfoldM(()) { _ =>
-      queue.take
-        .map(a => Some((a, ())))
-        .foldCauseM(
-          cause =>
-            // Dequeueing from a shutdown queue will result in interruption,
-            // so use that to signal the stream's end.
-            if (cause.interrupted) ZIO.succeed(None)
-            else ZIO.halt(cause),
-          ZIO.succeed
+    ZStream[R, E, A] {
+      ZManaged.reserve(
+        Reservation(
+          UIO(
+            queue.take.catchAllCause(
+              c => queue.isShutdown.flatMap(down => if (down && c.interrupted) Pull.end else Pull.halt(c))
+            )
+          ),
+          _ => queue.shutdown
         )
+      )
+    }
+
+  /**
+   * Creates a stream from a [[zio.ZQueue]] of queue `take` operations
+   */
+  final def fromQueueTake[R, E, A](queue: ZQueue[_, _, R, E, _, Take[E, A]]): ZStream[R, E, A] =
```

Comment:
@iravid Did I understand your suggestion correctly? ☝️

Reply:
Yup! Although, I realize now that this is just `fromQueue(q).unTake`. Do you think this justifies an additional constructor?

Reply:
Nope! ok, undoing that one.